src/Adapter/MSTest.TestAdapter/Execution/UnitTestRunner.cs

```diff
@@ -134,16 +136,55 @@ internal UnitTestResult[] RunSingleTest(TestMethod testMethod, IDictionary<strin
 
             if (!IsTestMethodRunnable(testMethod, testMethodInfo, out var notRunnableResult))
             {
+                bool shouldRunClassCleanup = false;
+                bool shouldRunClassAndAssemblyCleanup = false;
                 if (_classCleanupManager is null)
                 {
                     return notRunnableResult;
                 }
 
                 DebugEx.Assert(testMethodInfo is not null, "testMethodInfo should not be null.");
-                _classCleanupManager.MarkTestComplete(testMethodInfo, testMethod, out bool shouldRunClassCleanup);
+                _classCleanupManager.MarkTestComplete(testMethodInfo, testMethod, out shouldRunClassCleanup, out shouldRunClassAndAssemblyCleanup);
```

Comment:
It seems that most of the code below is extremely similar to the one of `RunCleanupsIfNeeded`, could we merge both?